design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+external resource] using the `crossplane.io/external-name` annotation:
+
+* The name of the external resource is set to the value of the annotation.
+* If the annotation is absent it is set to the managed resource's name.
+* The annotation is propagated from a resource claim to any managed resource it
+  dynamically provisions.
+
+This final point may pose a problem when one resource claim may provision many
+managed resources. If a resource class set the `crossplane.io/external-name`
+annotation and referenced an `AggregateResourceClass` that aggregated two
+resource classes of the same kind, both dynamically provisioned resources would
+attempt to use the same external name. External names are unique at different
+scopes (global, project, region) for different resources so this will not be an
+issue all resources, but it may for some.
+
+This document proposes two new external naming annotations be introduced at the
+resource class level to help resource class authors avoid external name
+conflicts:
+
+* `crossplane.io/external-name-prefix` - A value prepended to the external name
+  of any managed resource dynamically provisioned using the class.
+* `crossplane.io/external-name-suffix` - A value appended to the external name
+  of any managed resource dynamically provisioned using the class.
+
+These annotations would only be used during dynamic provisioning when a resource
+claim specified the `crossplane.io/external-name` annotation, resulting in an
+external name of `{prefix}{name}{suffix}`
+
+### Connection Secrets
+
+Many managed resources publish _connection secrets_; Kubernetes `Secret`
+resources containing the potentially sensitive details required to connect to
+the managed resource such as its hostname, ports, and credentials. These secrets
+are typically created in a namespace that is not accessible to application
+operators. Resource claim authors may request that the connection secret of
+their bound managed resource be propagated to the claim's namespace by
+specifying its `writeConnectionSecretToRef` field.
+
+This document proposes that `AggregateResource` connection secrets contain the
+intersection of the connection secrets of the managed resources they aggregate. 
+For example if an `AggregateResource` matched two managed resources with the
+following secrets:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-a
+stringData:
+  username: cooluser
+  password: verysecure
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: secret-b
+stringData:
+  url: https://example.org
+```
+
+The `AggregateResource` connection secret would be:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: aggregate-secret
+stringData:
+  username: cooluser
+  password: verysecure
+  url: https://example.org
+```
+
+An undefined connection secret key will win in the case of a conflict. Conflicts
```

Comment:
> You put a GKEClusterClass and CloudSQLInstanceClass in the same AggregateResourceClass 

Actually, we'd allow this `AggregateResourceClass` to exist but one of the two matching resource classes would be ignored depending on what kind of claim attempted to use the class. If a `KubernetesCluster` claim bound to it, it would match only the "compatible" class (`GKEClusterClass`) and ignore the `CloudSQLInstanceClass`. The opposite would be true if a `MySQLInstance` bound to it.

This was not very obvious in my initial design (it was hidden down under the technical implementation) so I've pushed a commit to make it clearer.

Reply:
@negz How about when you bind 3 `GKECluster`s to 1 `KubernetesClaim`? All fields are simply the same.

Reply:
> Does this mean the behavior is undefined

It does - so I'm basically saying "some key will win, but we do not specify which one". Maybe a random one, or maybe the first one, but not an empty/nil one.

Reply:
> How about when you bind 3 GKEClusters to 1 KubernetesClaim? All fields are simply the same.

This would be possible, but do you think it would be a likely use case? An infrastructure operator configuring an aggregate such that when someone asks for a KubernetesCluster, they get three GKE control planes? If someone wanted to do this the keys would conflict, and they'd have to use the annotation to break the conflict.

Reply:
> An infrastructure operator configuring an aggregate such that when someone asks for a KubernetesCluster, they get three GKE control planes?

I can't really tell how likely it is, probably useful for folks who want to create identical clusters in different regions or a configuration for `GKECluster` and one `OnPremCluster` from private infra-stack. It just looks pretty easy to do, so we might want to cover it.

> If someone wanted to do this the keys would conflict, and they'd have to use the annotation to break the conflict.

I am not sure about this one as I stated above. The key names are so low level, they are referred in the pod definition of the application, it's hard to calculate that as application owner and be careful about it. It's not impossible to include a logic to calculate the keys in your stack controller or deployment script in case the application is scheduled to use an aggregate resource but I can't say it's a very good UX.

About different types having conflicting keys. If we can say `all resources that publish a connection secret has to have a corresponding claim type`, we'd be able to not worry about the conflicts of different types since they will have their own secret referred by the claim. Otherwise, I do see `endpoint` key conflicting frequently.